### ISSUE TITLE
Auth: auth.proxy does not work with outh and auto_login enabled

### ISSUE THREAD

**Issue by mrevutskyi** [grafana/grafana#72476]
### What happened?

We have two auth plugins enabled:

```ini
[auth]
login_cookie_name = grafana_session


[auth.proxy]
enabled = true
header_name = X-WEBAUTH-USER
header_property = username
auto_sign_up = true
sync_ttl = 60
whitelist = ::1,127.0.0.1
headers = Name:X-WEBAUTH-NAME Email:X-WEBAUTH-EMAIL Groups:X-WEBAUTH-GROUPS
enable_login_token = true

[auth.azuread]
name = Azure AD
enabled = true
allow_sign_up = false
client_id = ****
client_secret = ***
scopes = openid email profile
auth_url = ***
token_url = ***
allowed_domains =
allowed_groups =
auto_login = true
```

When we make a call to `curl -k https://localhost:3000/login -H "X-WEBAUTH-USER: username"`  we do not get session cookie, but get redirected to azure auth page

### What did you expect to happen?

When auth proxy header is passed, user user should be authenticated via auth proxy, instead of being redirected to oauth

### Did this work before?

haven't tried with previous versions

### How do we reproduce it?

1. enable auth.proxy with `enable_login_token = true`
2. enable any other oauth plugin with `auto_login = true`
3. try to receive session cookie with `curl -k https://localhost:3000/login -H "X-WEBAUTH-USER: username"`

### Is the bug inside a dashboard panel?

no

### Environment (with versions)?

Grafana: 10.0.1
OS: Linux
Browser: Chrome, curl

### Grafana platform?

A downloaded binary

**Comment by usmangt** [comment]
Hi @mrevutskyi,

Thanks for opening this issue. It might be interesting to see the Grafana logs.

Could you please change the default log level to debug mode? 

To do this go to the server where grafana is installed and edit the configuration file (`/etc/grafana/grafana.ini`)

There you will this section;

```
#################################### Logging ##########################
[log]
# Either "console", "file", "syslog". Default is console and  file
# Use space to separate multiple modes, e.g. "console file"
;mode = console file

# Either "debug", "info", "warn", "error", "critical", default is "info"
;level = info
```
Set the parameter level to debug and remove the colon in the beginning i.e.

`level = debug`

Then restart the grafana service and resent the logs once the issue occurs again.

**Comment by mrevutskyi** [comment]
Hi, here is the log
```
logger=auth.proxy t=2023-07-31T15:15:10.011970176-07:00 level=debug msg="Trying to log user in" username=mrevutsk ignoreCache=false
logger=auth.proxy t=2023-07-31T15:15:10.01202708-07:00 level=debug msg="Getting user ID via auth cache" cacheKey=auth-proxy-sync-ttl:54ac725a1d659b10e0dce65dac468f62
logger=auth.proxy t=2023-07-31T15:15:10.012756037-07:00 level=debug msg="Successfully got user ID via auth cache" id=13
logger=auth.proxy t=2023-07-31T15:15:10.01279944-07:00 level=debug msg="Got user ID, getting full user info" userID=13
logger=auth.proxy t=2023-07-31T15:15:10.014270654-07:00 level=debug msg="Successfully got user info" userID=13 username=mrevutsk
logger=context userId=13 orgId=1 uname=mrevutsk t=2023-07-31T15:15:10.014911603-07:00 level=debug msg="Updating last user_seen_at" user_id=13
logger=token t=2023-07-31T15:15:10.022115159-07:00 level=debug msg=FeatureEnabled feature=accesscontrol.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=accesscontrol t=2023-07-31T15:15:10.022176864-07:00 level=debug msg="fetch permissions from store" key=rbac-permissions-1-user-13
logger=accesscontrol t=2023-07-31T15:15:10.024349532-07:00 level=debug msg="cache permissions" key=rbac-permissions-1-user-13
logger=token t=2023-07-31T15:15:10.024496743-07:00 level=debug msg=FeatureEnabled feature=caching.queries enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027071442-07:00 level=debug msg=FeatureEnabled feature=dspermissions.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027126147-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:autometrics
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027145648-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:informed
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027157849-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:jira
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.02717005-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:yE4i8fyVz
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027193552-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:P83C9889B5738EA36
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027205553-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:d16190ee-8a04-493b-8191-732ba3443abd
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027227454-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:PB48DB0ED74D5C099
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027239655-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:P3E65445149F74386
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027251456-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:mysql-makto-db-015-osua_observer
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027275958-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:PE32D5DB5A492A8C9
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.027287459-07:00 level=debug msg="matched scope" userscope=datasources:* targetscope=datasources:uid:mrevutsk1
logger=token t=2023-07-31T15:15:10.027449472-07:00 level=debug msg=FeatureEnabled feature=accesscontrol enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027479874-07:00 level=debug msg=FeatureEnabled feature=accesscontrol.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027494775-07:00 level=debug msg=FeatureEnabled feature=admin enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027520577-07:00 level=debug msg=FeatureEnabled feature=analytics enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027534878-07:00 level=debug msg=FeatureEnabled feature=analytics.writers enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027549179-07:00 level=debug msg=FeatureEnabled feature=auditing enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027573681-07:00 level=debug msg=FeatureEnabled feature=caching enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027603083-07:00 level=debug msg=FeatureEnabled feature=caching.api enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027622585-07:00 level=debug msg=FeatureEnabled feature=caching.queries enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027669589-07:00 level=debug msg=FeatureEnabled feature=caching.resources enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027706191-07:00 level=debug msg=FeatureEnabled feature=config.vault enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027729593-07:00 level=debug msg=FeatureEnabled feature=dspermissions enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027753895-07:00 level=debug msg=FeatureEnabled feature=dspermissions.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027794098-07:00 level=debug msg=FeatureEnabled feature=encryption.aesgcm enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027850502-07:00 level=debug msg=FeatureEnabled feature=enterprise.plugins enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027879305-07:00 level=debug msg=FeatureEnabled feature=grafanacloud enabled=false licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027895106-07:00 level=debug msg=FeatureEnabled feature=kms.encryption enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027919708-07:00 level=debug msg=FeatureEnabled feature=ldapdebug enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027958711-07:00 level=debug msg=FeatureEnabled feature=ldapsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.027995314-07:00 level=debug msg=FeatureEnabled feature=publicDashboardsEmailSharing enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028010315-07:00 level=debug msg=FeatureEnabled feature=provisioning enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028038917-07:00 level=debug msg=FeatureEnabled feature=recordedqueries enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028053418-07:00 level=debug msg=FeatureEnabled feature=reports enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.02807722-07:00 level=debug msg=FeatureEnabled feature=reports.creation enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028091921-07:00 level=debug msg=FeatureEnabled feature=reports.email enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028115923-07:00 level=debug msg=FeatureEnabled feature=reports.pdf enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028130124-07:00 level=debug msg=FeatureEnabled feature=saml enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028161126-07:00 level=debug msg=FeatureEnabled feature=teamsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028178828-07:00 level=debug msg=FeatureEnabled feature=teamgroupsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.02820373-07:00 level=debug msg=FeatureEnabled feature=userlimits enabled=false licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.028221131-07:00 level=debug msg=FeatureEnabled feature=whitelabeling enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.029532232-07:00 level=debug msg=FeatureEnabled feature=saml enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.030101776-07:00 level=debug msg="matched scope" userscope=settings:* targetscope=settings:auth.saml:*
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.030140879-07:00 level=debug msg="matched scope" userscope=settings:* targetscope=settings:auth.saml:*
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.030158481-07:00 level=debug msg="matched scope" userscope=settings:* targetscope=settings:*
logger=token t=2023-07-31T15:15:10.030192383-07:00 level=debug msg=FeatureEnabled feature=accesscontrol.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=accesscontrol t=2023-07-31T15:15:10.030214185-07:00 level=debug msg="fetch permissions from store" key=rbac-permissions-0-user-13
logger=accesscontrol t=2023-07-31T15:15:10.03131357-07:00 level=debug msg="cache permissions" key=rbac-permissions-0-user-13
logger=accesscontrol.evaluator t=2023-07-31T15:15:10.031393976-07:00 level=debug msg="matched scope" userscope=settings:* targetscope=settings:*
logger=token t=2023-07-31T15:15:10.031998523-07:00 level=debug msg=FeatureEnabled feature=accesscontrol.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=config.storage t=2023-07-31T15:15:10.032119232-07:00 level=debug msg="looking for datasource cache list in local cache"
logger=config.storage t=2023-07-31T15:15:10.032143834-07:00 level=debug msg="datasource cache config list not found in cache"
logger=config.storage t=2023-07-31T15:15:10.032158135-07:00 level=debug msg="retrieving all datasource cache configs"
logger=token t=2023-07-31T15:15:10.032509662-07:00 level=debug msg=FeatureEnabled feature=reports enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032576468-07:00 level=debug msg=FeatureEnabled feature=accesscontrol enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.03260537-07:00 level=debug msg=FeatureEnabled feature=accesscontrol.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032621171-07:00 level=debug msg=FeatureEnabled feature=admin enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032731579-07:00 level=debug msg=FeatureEnabled feature=analytics enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032761982-07:00 level=debug msg=FeatureEnabled feature=analytics.writers enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032776983-07:00 level=debug msg=FeatureEnabled feature=auditing enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032805885-07:00 level=debug msg=FeatureEnabled feature=caching enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032821586-07:00 level=debug msg=FeatureEnabled feature=caching.api enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032846588-07:00 level=debug msg=FeatureEnabled feature=caching.queries enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.03286559-07:00 level=debug msg=FeatureEnabled feature=caching.resources enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032891092-07:00 level=debug msg=FeatureEnabled feature=config.vault enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032905593-07:00 level=debug msg=FeatureEnabled feature=dspermissions enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032930695-07:00 level=debug msg=FeatureEnabled feature=dspermissions.enforcement enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.032945696-07:00 level=debug msg=FeatureEnabled feature=encryption.aesgcm enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.0329964-07:00 level=debug msg=FeatureEnabled feature=enterprise.plugins enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033027502-07:00 level=debug msg=FeatureEnabled feature=grafanacloud enabled=false licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033044304-07:00 level=debug msg=FeatureEnabled feature=kms.encryption enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033067805-07:00 level=debug msg=FeatureEnabled feature=ldapdebug enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033083307-07:00 level=debug msg=FeatureEnabled feature=ldapsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033107209-07:00 level=debug msg=FeatureEnabled feature=publicDashboardsEmailSharing enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.03312241-07:00 level=debug msg=FeatureEnabled feature=provisioning enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033136611-07:00 level=debug msg=FeatureEnabled feature=recordedqueries enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033162013-07:00 level=debug msg=FeatureEnabled feature=reports enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033179014-07:00 level=debug msg=FeatureEnabled feature=reports.creation enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033203016-07:00 level=debug msg=FeatureEnabled feature=reports.email enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033217617-07:00 level=debug msg=FeatureEnabled feature=reports.pdf enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033234718-07:00 level=debug msg=FeatureEnabled feature=saml enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.03325972-07:00 level=debug msg=FeatureEnabled feature=teamsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033277122-07:00 level=debug msg=FeatureEnabled feature=teamgroupsync enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033307924-07:00 level=debug msg=FeatureEnabled feature=userlimits enabled=false licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033353027-07:00 level=debug msg=FeatureEnabled feature=whitelabeling enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=token t=2023-07-31T15:15:10.033375029-07:00 level=debug msg=FeatureEnabled feature=whitelabeling enabled=true licenseStatus=Valid hasLicense=true hasValidLicense=true products=[grafana-enterprise]
logger=context userId=13 orgId=1 uname=mrevutsk t=2023-07-31T15:15:10.033434034-07:00 level=info msg="OAuth auto login enabled. Redirecting to /g/login/azuread"
logger=context userId=13 orgId=1 uname=mrevutsk t=2023-07-31T15:15:10.033500139-07:00 level=info msg="Request Completed" method=HEAD path=/login status=307 remote_addr=[::1] time_ms=21 duration=21.564466ms size=0 referer= handler=/login
```

Please also take a look at the linked PR #72477 , hopefully it will help to understand the issue better

**Comment by kalleep** [comment]
Hey, the auto_login feature does not play well when using multiple providers. 

Ideally it would be disabled if using more than one auth provider, but seems like we only disable it if more than one provider has `auto_login = enabled` https://github.com/grafana/grafana/blob/main/pkg/api/login.go#L159-L167.

I am not sure this is a use-case we want to support but will bring it up in our next team meeting.

**Comment by mrevutskyi** [comment]
We use a combination of `auth.azuread` and `auth.proxy`. Where we verify credentials using `auth.azuread` but then we have additional business logic before we issue a session cookie using `auth.proxy`, so this use case is very important to us.

Thank you.

**Comment by kalleep** [comment]
Took a look at your fix and that one seems reasonable.

**Comment by kalleep** [comment]
That is strange. Looking at the commits it seems like the email is not linked to your github account.

**Comment by mrevutskyi** [comment]
Thank you for reviewing and merging the PR. 
When should I expect it to get released?

**Comment by kalleep** [comment]
No worries. Initially planed to go out with 10.2 but decided to backport the fix to the next 10.1.x release
